cmd/otel-allocator/target/discovery.go

```diff
@@ -105,43 +127,122 @@ func (m *Discoverer) ApplyConfig(source allocatorWatcher.EventSource, scrapeConf
 	return m.manager.ApplyConfig(discoveryCfg)
 }
 
-func (m *Discoverer) Watch(fn func(targets map[string]*Item)) error {
-	labelsBuilder := labels.NewBuilder(labels.EmptyLabels())
+func (m *Discoverer) Run() error {
+	err := m.run(m.manager.SyncCh())
+	if err != nil {
+		m.log.Error(err, "Service Discovery watch event failed")
+		return err
+	}
+	<-m.close
+	m.log.Info("Service Discovery watch event stopped: discovery manager closed")
+	return nil
+}
+
+// UpdateTsets updates the target sets to be scraped.
+func (m *Discoverer) UpdateTsets(tsets map[string][]*targetgroup.Group) {
+	m.mtxScrape.Lock()
+	m.targetSets = tsets
+	m.mtxScrape.Unlock()
```

Comment:
Do we even need to lock here? We only ever assign to this attribute, never modify the value.

Reply:
I'm worried if we moved this log we can update a map is being used by the Reload func.
wdyt @swiatekm ?

Reply:
We never modify the map, we just change what the struct field is pointing to, so I think this is fine. In `Reload` we only reference the value of this map once too.

Reply:
Sorry to go back and forth on this, but it seems like we actually do need the mutex. Go only guarantees atomic writes for single machine word values, so this is actually a race condition without the lock, albeit an unlikely one.

What is somewhat alarming is that the race detector doesn't cause tests to fail.

Reply:
There is nothing to apologize for, @swiatekm, and I wasn't sure as well about it.
But since the race detector didn't fail the tests, I was fine 😅

Reply:
Maybe we don't have a test where the updates are concurrent? Could do that by putting multiple values into the Discoverer's input channel.

Reply:
We can do that in a follow-up though.